File: src/services/task-service.ts
import ApiClient from './api-client';
import { Task } from '@/types/task.types';

const apiClient = new ApiClient();

export const taskService = {
  // Get all tasks for a user
  async getTasks(userId: string): Promise<Task[]> {
    try {
      const response = await apiClient.get<{ success: boolean; tasks: Task[] }>(`/api/users/${userId}/tasks`);
      return response.data.success ? response.data.tasks || [] : [];
    } catch (error) {
      console.error('Failed to fetch tasks:', error);
      throw error;
    }
  },

  // Create a new task
  async createTask(taskData: Omit<Task, 'id' | 'createdAt' | 'updatedAt' | 'completedAt'>): Promise<Task> {
    try {
      const response = await apiClient.post<{ success: boolean; task: Task}>(
        `/api/users/${taskData.userId}/tasks`,
        {
          title: taskData.title,
          description: taskData.description || "",
          completed: taskData.completed
        }
      );
      if (response.data.success) {
        return response.data.task;
      } else {
        throw new Error(response.data.task ? "Task creation failed" : "Unknown error");
      }
    } catch (error) {
      console.error('Failed to create task:', error);
      throw error;
    }
  },

  // Update an existing task
  async updateTask(userId: string, taskId: string, taskData: Partial<Omit<Task, 'id' | 'userId' | 'createdAt' | 'updatedAt' | 'completedAt'>>): Promise<Task> {
    try {
      const response = await apiClient.put<{ success: boolean; task: Task}>(
        `/api/users/${userId}/tasks/${taskId}`,
        taskData
      );
      if (response.data.success) {
        return response.data.task;
      } else {
        throw new Error("Task update failed");
      }
    } catch (error) {
      console.error('Failed to update task:', error);
      throw error;
    }
  },

  // Delete a task
  async deleteTask(userId: string, taskId: string): Promise<void> {
    try {
      const response = await apiClient.delete<{ success: boolean; message?: string }>(`/api/users/${userId}/tasks/${taskId}`);
      if (!response.data.success) {
        throw new Error(response.data.message || "Task deletion failed");
      }
    } catch (error) {
      console.error('Failed to delete task:', error);
      throw error;
    }
  },

  // Toggle task completion status
  async toggleTaskCompletion(userId: string, taskId: string, completed: boolean): Promise<Task> {
    try {
      const response = await apiClient.patch<{ success: boolean; task: Task}>(
        `/api/users/${userId}/tasks/${taskId}/complete`,
        { completed }
      );
      if (response.data.success) {
        return response.data.task;
      } else {
        throw new Error("Task completion toggle failed");
      }
    } catch (error) {
      console.error('Failed to toggle task completion:', error);
      throw error;
    }
  }
};

File: backend/src/services/task_service.py
from sqlmodel import Session, select
from typing import List, Optional
from ..models.task import Task, TaskCreate, TaskUpdate, TaskResponse
from fastapi import HTTPException, status
from datetime import datetime


class TaskService:
    @staticmethod
    def get_all_tasks(session: Session, user_id: int, completed: Optional[bool] = None) -> List[Task]:
        """
        Get all tasks for a specific user, optionally filtered by completion status
        """
        statement = select(Task).where(Task.user_id == user_id)

        if completed is not None:
            statement = statement.where(Task.completed == completed)

        tasks = session.exec(statement).all()
        return tasks

    @staticmethod
    def get_task_by_id(session: Session, user_id: int, task_id: int) -> Task:
        """
        Get a specific task by ID for a user
        """
        statement = select(Task).where(Task.id == task_id, Task.user_id == user_id)
        task = session.exec(statement).first()

        if not task:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Task not found"
            )

        return task

    @staticmethod
    def create_task(session: Session, task_create: TaskCreate) -> Task:
        """
        Create a new task
        """
        # Create the task instance from the input data
        db_task = Task(
            title=task_create.title,
            description=task_create.description,
            completed=task_create.completed,
            user_id=task_create.user_id,
        )

        # Add the task to the session
        session.add(db_task)
        session.commit()
        session.refresh(db_task)

        return db_task

    @staticmethod
    def update_task(session: Session, user_id: int, task_id: int, task_update: TaskUpdate) -> Task:
        """
        Update an existing task
        """
        statement = select(Task).where(Task.id == task_id, Task.user_id == user_id)
        db_task = session.exec(statement).first()

        if not db_task:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Task not found"
            )

        # Update only the fields that are provided
        update_data = task_update.dict(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_task, field, value)

        # Update the updated_at timestamp
        db_task.updated_at = datetime.utcnow()

        session.add(db_task)
        session.commit()
        session.refresh(db_task)

        return db_task

    @staticmethod
    def delete_task(session: Session, user_id: int, task_id: int) -> bool:
        """
        Delete a task by ID for a user
        """
        statement = select(Task).where(Task.id == task_id, Task.user_id == user_id)
        db_task = session.exec(statement).first()

        if not db_task:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Task not found"
            )

        session.delete(db_task)
        session.commit()

        return True

    @staticmethod
    def toggle_task_completion(session: Session, user_id: int, task_id: int, completed: bool) -> Task:
        """
        Toggle the completion status of a task
        """
        statement = select(Task).where(Task.id == task_id, Task.user_id == user_id)
        db_task = session.exec(statement).first()

        if not db_task:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Task not found"
            )

        db_task.completed = completed
        db_task.updated_at = datetime.utcnow()

        session.add(db_task)
        session.commit()
        session.refresh(db_task)

        return db_task